pkg/scale/scale.go

```diff
@@ -15,497 +15,377 @@
 		}
+		var bID int
+		bID, err = strconv.Atoi(brokerID)
+		if err != nil {
+			cc.log.Error(err, "failed to cast broker ID from string to integer", "broker_id", brokerID)
+			return nil, err
+		}
+		brokersToRemove = append(brokersToRemove, int32(bID))
 	}
 
-	err = json.Unmarshal(body, &response)
-	if err != nil {
-		return brokerWithLeastPartition, err
+	if len(brokersToRemove) == 0 {
+		return &Result{
+			State: v1beta1.CruiseControlTaskCompleted,
+		}, nil
 	}
 
-	replicaCountByBroker := response.KafkaBrokerState.ReplicaCountByBrokerID
-	replicaCount := float64(99999)
-
-	for brokerID, replica := range replicaCountByBroker {
-		if replicaCount > replica {
-			replicaCount = replica
-			brokerWithLeastPartition = brokerID
-		}
+	rmBrokerReq := &api.RemoveBrokerRequest{
+		AllowCapacityEstimation: true,
+		BrokerIDs:               brokersToRemove,
+		DataFrom:                types.ProposalDataSourceValidWindows,
+		UseReadyDefaultGoals:    true,
 	}
-	return brokerWithLeastPartition, nil
+	rmBrokerResp, err := cc.client.RemoveBroker(rmBrokerReq)
+	if err != nil {
+		return &Result{
+			TaskID:    rmBrokerResp.TaskID,
+			StartedAt: rmBrokerResp.Date,
+			State:     v1beta1.CruiseControlTaskCompletedWithError,
+			Err:       fmt.Sprintf("%v", err),
+		}, err
+	}
+
+	return &Result{
+		TaskID:    rmBrokerResp.TaskID,
+		StartedAt: rmBrokerResp.Date,
+		State:     v1beta1.CruiseControlTaskActive,
+	}, nil
 }
 
-// UpScaleCluster upscales Kafka cluster
-func (cc *cruiseControlScaler) UpScaleCluster(brokerIDs []string) (string, string, error) {
-	liveBrokers, err := cc.GetLiveKafkaBrokersFromCruiseControl(brokerIDs)
+// RebalanceDisks performs a disk rebalance via Cruise Control for the provided list of brokers.
+func (cc *cruiseControlScaler) RebalanceDisks(brokerIDs ...string) (*Result, error) {
+	clusterLoadResp, err := cc.client.KafkaClusterLoad(api.KafkaClusterLoadRequestWithDefaults())
 	if err != nil {
-		return "", "", err
+		return nil, err
 	}
-	ready := bcutil.AreStringSlicesIdentical(liveBrokers, brokerIDs)
 
-	if !ready {
-		return "", "", errors.New("broker(s) not yet ready in cruise-control")
-	}
+	brokerIDsMap := stringSliceToMap(brokerIDs)
 
-	options := map[string]string{
-		"json":     "true",
-		"dryrun":   "false",
-		"brokerid": strings.Join(brokerIDs, ","),
+	brokersWithEmptyDisks := make([]int32, 0, len(brokerIDs))
+	for _, brokerStat := range clusterLoadResp.Result.Brokers {
+		if _, ok := brokerIDsMap[strconv.Itoa(int(brokerStat.Broker))]; !ok {
+			continue
+		}
+		for _, diskState := range brokerStat.DiskState {
+			if diskState.NumReplicas <= 0 {
+				brokersWithEmptyDisks = append(brokersWithEmptyDisks, brokerStat.Broker)
+			}
+		}
 	}
 
-	uResp, err := cc.postCruiseControl(addBrokerAction, options)
-	if err != nil && err != errCruiseControlNotReturned200 {
-		log.Error(err, "can't upscale cluster gracefully since post to cruise-control failed")
-		return "", "", err
+	if len(brokersWithEmptyDisks) == 0 {
```

Comment:
When this method is called for a broker X with a single disk attached (log dir) which is empty (e.g. some custom message partitioning logic is used by producers which lead to message replicas to be stored only on the other brokers) a disk rebalance operation will triggered in CC unnecessarily since this condition won't be satisfied.

Could this cause a loop in the reconcile logic? or not as the triggered task is being tracked and once completed this method won't be invoked for broker X any more (unless a new disk is added again)

Reply:
> When this method is called for a broker X with a single disk attached (log dir) which is empty (e.g. some custom message partitioning logic is used by producers which lead to message replicas to be stored only on the other brokers) a disk rebalance operation will triggered in CC unnecessarily since this condition won't be satisfied.

The code checks whether topic partition replicas are assigned to the LogDir/Disk or not. This is independent from what is the partition key used by Producers pushing messages to the topic. as Kafka assignes partition replicas to brokers (and LogDirs) when the Topic is created.

Disk rebalance is only triggered if there are one or more empty LogDirs (have no partition replicas assigned) for Broker(s) identified by `brokerIDs`.

> Could this cause a loop in the reconcile logic? or not as the triggered task is being tracked and once completed this method won't be invoked for broker X any more (unless a new disk is added again)

The latter. The task is tracked by the controller and it does not trigger a new disk rebalance action unless the task has gone from Cruise Control (it has been restarted since then).

I'll update the PR to add one tiny improvement to this logic which makes sure that we don't trigger disk rebalance for dead disks.

Reply:
Added the check for dead disks here: https://github.com/banzaicloud/koperator/pull/749/commits/71d4522248e505258c965fea06e2726fa7284e41